File: apps/api/app/routes/tags.py
from __future__ import annotations

from fastapi import APIRouter, Depends, Query, Response, status
from sqlalchemy.ext.asyncio import AsyncSession

from app.auth.dependencies import get_current_user
from app.db.session import get_db
from app.models.user import User
from app.schemas.tag import (
    PruneOrphansResult,
    TagMerge,
    TagMergeResult,
    TagRead,
    TagReadWithCount,
    TagRename,
)
from app.services import tags_service

router = APIRouter(prefix="/api/tags", tags=["tags"])


@router.get("", response_model=list[TagRead])
async def list_tags_route(
    q: str | None = None,
    user: User = Depends(get_current_user),
    session: AsyncSession = Depends(get_db),
) -> list[TagRead]:
    tags = await tags_service.list_tags(session, user.id, q)
    return [TagRead.model_validate(t) for t in tags]


@router.get("/with-counts", response_model=list[TagReadWithCount])
async def list_tags_with_counts_route(
    user: User = Depends(get_current_user),
    session: AsyncSession = Depends(get_db),
) -> list[TagReadWithCount]:
    rows = await tags_service.list_with_counts(session, user.id)
    return [TagReadWithCount(**r) for r in rows]


@router.patch("/{tag_id}", response_model=TagRead)
async def rename_tag_route(
    tag_id: int,
    body: TagRename,
    user: User = Depends(get_current_user),
    session: AsyncSession = Depends(get_db),
) -> TagRead:
    tag = await tags_service.rename_tag(session, user.id, tag_id, body.name)
    return TagRead.model_validate(tag)


@router.post("/{tag_id}/merge", response_model=TagMergeResult)
async def merge_tag_route(
    tag_id: int,
    body: TagMerge,
    user: User = Depends(get_current_user),
    session: AsyncSession = Depends(get_db),
) -> TagMergeResult:
    result = await tags_service.merge_tags(session, user.id, tag_id, body.target_id)
    return TagMergeResult(**result)


@router.delete(
    "/{tag_id}", status_code=status.HTTP_204_NO_CONTENT, response_class=Response
)
async def delete_tag_route(
    tag_id: int,
    force: bool = Query(default=False),
    user: User = Depends(get_current_user),
    session: AsyncSession = Depends(get_db),
) -> Response:
    await tags_service.delete_tag(session, user.id, tag_id, force=force)
    return Response(status_code=status.HTTP_204_NO_CONTENT)


@router.post("/prune-orphans", response_model=PruneOrphansResult)
async def prune_orphans_route(
    user: User = Depends(get_current_user),
    session: AsyncSession = Depends(get_db),
) -> PruneOrphansResult:
    count = await tags_service.prune_orphans(session, user.id)
    return PruneOrphansResult(deleted_count=count)

File: apps/web/lib/api/tags.ts
import type { paths } from "@ims/api-types"
import { apiFetch } from "./client"

type TagRead = paths["/api/tags"]["get"]["responses"]["200"]["content"]["application/json"][number]
type TagReadWithCount = paths["/api/tags/with-counts"]["get"]["responses"]["200"]["content"]["application/json"][number]
type TagMergeResult = paths["/api/tags/{tag_id}/merge"]["post"]["responses"]["200"]["content"]["application/json"]
type PruneOrphansResult = paths["/api/tags/prune-orphans"]["post"]["responses"]["200"]["content"]["application/json"]

export type { TagRead, TagReadWithCount, TagMergeResult, PruneOrphansResult }

export async function listTags(
  q: string | undefined,
  accessToken: string | null,
): Promise<TagRead[]> {
  const qs = q ? `?q=${encodeURIComponent(q)}` : ""
  const res = await apiFetch(`/tags${qs}`, { accessToken })
  return (await res.json()) as TagRead[]
}

export async function listTagsWithCounts(
  accessToken: string | null,
): Promise<TagReadWithCount[]> {
  const res = await apiFetch("/tags/with-counts", { accessToken })
  return (await res.json()) as TagReadWithCount[]
}

export async function renameTag(
  tagId: number,
  name: string,
  accessToken: string | null,
): Promise<TagRead> {
  const res = await apiFetch(`/tags/${tagId}`, {
    accessToken,
    method: "PATCH",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify({ name }),
  })
  return (await res.json()) as TagRead
}

export async function mergeTag(
  sourceId: number,
  targetId: number,
  accessToken: string | null,
): Promise<TagMergeResult> {
  const res = await apiFetch(`/tags/${sourceId}/merge`, {
    accessToken,
    method: "POST",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify({ target_id: targetId }),
  })
  return (await res.json()) as TagMergeResult
}

export async function deleteTag(
  tagId: number,
  force: boolean,
  accessToken: string | null,
): Promise<void> {
  const qs = force ? "?force=true" : ""
  await apiFetch(`/tags/${tagId}${qs}`, { accessToken, method: "DELETE" })
}

export async function pruneOrphanTags(
  accessToken: string | null,
): Promise<PruneOrphansResult> {
  const res = await apiFetch("/tags/prune-orphans", {
    accessToken,
    method: "POST",
  })
  return (await res.json()) as PruneOrphansResult
}
